src/utils/slpParser.ts

```diff
@@ -0,0 +1,114 @@
+        return this.followerFrames;
+    }
+
+    public handleGameEnd(payload: GameEndType): void {
+        payload = payload as GameEndType;
+        this.gameEnd = payload;
+    }
+
+    public getStats(): StatsType {
+        return {
+            ...this.statsComputer.getStats(),
+            gameComplete: this.gameEnd !== null,
+        };
+    }
+
+    public handleGameStart(payload: GameStartType): void {
+        if (!payload.stageId) {
+            return;
+        }
+
+        this.settings = payload;
+        const players = payload.players;
+        this.settings.players = players.filter(player => player.type !== 3);
+        this.playerIndices = getSinglesOpponentIndicesFromSettings(this.settings);
+        this.statsComputer = new Stats(this.playerIndices);
+    }
+
+    public handlePostFrameUpdate(payload: PostFrameUpdateType): void {
+        if (payload.frame === null) {
+            // Once we are an frame -122 or higher we are done getting match settings
+            // Tell the iterator to stop
+            return;
+        }
+        // handle settings calculation
+        if (payload.frame <= Frames.FIRST) {
+            const playerIndex = payload.playerIndex;
+            const playersByIndex = _.keyBy(this.settings.players, 'playerIndex');
+
+            switch (payload.internalCharacterId) {
+                case 0x7:
+                    playersByIndex[playerIndex].characterId = 0x13; // Sheik
+                    break;
+                case 0x13:
+                    playersByIndex[playerIndex].characterId = 0x12; // Zelda
+                    break;
+            }
+        }
+    }
+
+    public getLatestFrame(): FrameEntryType | null {
+        // return this.playerFrames[this.latestFrameIndex];
+
+        // TODO: Modify this to check if we actually have all the latest frame data and return that
+        // TODO: If we do. For now I'm just going to take a shortcut
+        const allFrames = this.getFrames();
+        const frameIndex = this.latestFrameIndex || Frames.FIRST;
+        const indexToUse = this.gameEnd ? frameIndex : frameIndex - 1;
+        return _.get(allFrames, indexToUse) || null;
+    }
+
+    public handleFrameUpdate(command: Command, payload: PreFrameUpdateType | PostFrameUpdateType): FrameEntryType {
+        payload = payload as PostFrameUpdateType;
+        if (!payload.frame && payload.frame !== 0) {
+            // If payload is messed up, stop iterating. This shouldn't ever happen
+            return;
+        }
+
+        const location = command === Command.PRE_FRAME_UPDATE ? "pre" : "post";
+        const frames = payload.isFollower ? this.followerFrames : this.playerFrames;
+        this.latestFrameIndex = payload.frame;
+        _.set(frames, [payload.frame, 'players', payload.playerIndex, location], payload);
+        _.set(frames, [payload.frame, 'frame'], payload.frame);
+
+        const frame = frames[payload.frame];
+        this.statsComputer.processFrame(frame);
```

Comment:
My understanding is that your method computes stats starting from the _process function. You're effectively hooking into the logic that makes iterateEvents not re-compute pre-computed data in order to make your stats stuff also not re-compute what has already been computed. Couple issues:

For files that are already complete, it no longer seems possible to call getSettings and have the lib only read the game settings section and nothing else. Even if all you need are settings, the lib will process the entire file regardless. This is highly problematic for the desktop app which can load thousands of replays at once to populate a table of games. This was... somewhat fast before because it only processed the game start payload from each file. Now it will parse everything and will likely be much much slower
Stats are now always computed even if you don't want them. Sometimes I might only want frame data and not subject myself to the performance loss of computing stats, this might be the case if I'm making a HUD replacement software.

Reply:
Here are my (rough) thoughts on the two:
1. It might be possible to solve this by adding a simple arg to `_process` to allow it to short circuit on a getSettings call?
2. This should still be possible to tie to `getStats` but you'd have to manage the state for what the last frame that has been processed through the stats engine was externally somewhere. Basically every time `getStats` is called it iterates through all the frames that are new since the last `getStats` call?

Reply:
Yep I've added lazy processing of frames which is on by default. But I wanted the option to process stats on the fly as each frame is added so there's an option to enable that.